### PR TITLE
Generate test coverage report

### LINKED ISSUE #438: Generate test coverage report
Generate test coverage using Jacoco plugin. 
Create a new workflow which generates test coverage report.
Add a badge in the readme file which shows the current coverage percentage.

### PR DESCRIPTION
**Pull request by wanglin100**
fix #438

### CODE REVIEW

**Review by swaroopar — CHANGES_REQUESTED** [review]
@wanglin100 Thank you for adding this. It was much needed.
1. would be good to have an issue to track this. 
2. For me,  the requirement is to have another workflow which generates this report. 
3. To have a badge in the readme which shows the current coverage percentage.

File: pom.xml [hunk 1]
```diff
@@ -122,6 +122,18 @@
         </profile>
     </profiles>
     <dependencies>
+        <dependency>
+            <groupId>org.jacoco</groupId>
+            <artifactId>org.jacoco.core</artifactId>
+            <version>0.8.10</version>
```
Comment (baixinsui) [inline pom.xml:128]: Define version variables instead of fixed values

File: pom.xml [hunk 2]
```diff
@@ -239,6 +251,30 @@
...
+                        </goals>
+                    </execution>
+                    <execution>
+                        <id>report</id>
+                        <phase>prepare-package</phase>
+                        <goals>
+                            <goal>report</goal>
+                        </goals>
+                    </execution>
+                </executions>
```
Comment (baixinsui) [inline pom.xml:272]: We can use this plugin in the pom.xml of runtime to execute a report-aggregate task which can aggregate the test coverage of all modules.

File: pom.xml (hunk 1 above)
Comment (wanglin100) [reply pom.xml]: ok

File: pom.xml (hunk 2 above)
Comment (wanglin100) [reply pom.xml:274]: ok,I have modified

**Review by baixinsui — APPROVED** [review]
Looks good to me.

**Comment by iskey** [comment]
I think we must involve the `gist` If we want a dynamic badge coverage workflow.

File: .github/workflows/coverage.yml [hunk 3]
```diff
@@ -0,0 +1,46 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Huawei Inc.
+#
+name: Generate Coverage Badge
```
Comment (swaroopar) [inline .github/workflows/coverage.yml:5]: pls change the name to - `xpanse-test-coverage`

File: .github/workflows/coverage.yml [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+          coverage=$(grep -oP '<td class="ctr2">\K(\d+)%' target/coverage-reports/index.html | head -n 1)
+          coverage=${coverage%\%}
+          echo "COVERAGE_PERCENTAGE=$coverage" >> $GITHUB_ENV
+
+      - name: Create Coverage Badge
+        uses: schneegans/dynamic-badges-action@v1.6.0
+        with:
+          auth: ${{ secrets.GIST_TOKEN }}
+          gistID: 3d9c022b98734fbf615c21136abe4add
+          filename: gistfile1.txt
```
Comment (swaroopar) [inline .github/workflows/coverage.yml:41]: need better name. `coverage.txt` ?

Comment (wanglin100) [reply .github/workflows/coverage.yml:41]: name

 coverage.txt is fine,little effect here

File: .github/workflows/coverage.yml (hunk 3 above)
Comment (wanglin100) [reply .github/workflows/coverage.yml:5]: ok

File: .github/workflows/coverage.yml [hunk 5]
```diff
@@ -0,0 +1,46 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Huawei Inc.
+#
+name: xpanse-test-coverage
+
+on:
+  push:
+    branches: [main]
+  pull_request:
```
Comment (swaroopar) [inline .github/workflows/coverage.yml:10]: Not needed. the report must be generated only on pushes to main branch.

File: .github/workflows/coverage.yml [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    branches: [main]
+
+jobs:
```
Comment (swaroopar) [inline .github/workflows/coverage.yml:13]: we must also ensure that the job runs only on actual repo and not on forked repos.
Please add a check like this -> https://github.com/eclipse-xpanse/xpanse/blob/main/.github/workflows/release.yml#L26

**Review by swaroopar — CHANGES_REQUESTED** [review]
new comments added.

**Comment by swaroopar** [comment]
same code was merged with a different PR.

**Pull request closed without merging**